New posting of posparams fitting code

joesilber opened the pull request:
This is posparams code as-is from prototyping outside desimeter. Minor adjustments to make paths work with the desimeter directory structure. Plenty more to do, but does fully work on my desktop Windows machine. C.f. DESI-5691-v1 for more context and typical output files.

sbailey commented:
Note that this is our first contribution from someone on a Windows machine so we'll want to double check for any gotchas about CR/LF, forward slash vs. backslash file paths, etc.

joesilber commented:
Good point. If needed I will develop on linux, rather than have any special
handling just for me.

On Mon, Apr 13, 2020, 6:20 PM Stephen Bailey <notifications@github.com>
wrote:

> Note that this is our first contribution from someone on a Windows machine
> so we'll want to double check for any gotchas about CR/LF, forward slash
> vs. backslash file paths, etc.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/desihub/desimeter/pull/54#issuecomment-613174596>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AN724Q27S5FWWZN7I7IH5N3RMO25NANCNFSM4MHLI54A>
> .
>

sbailey commented on bin/fit_posparams:
```diff
@@ -0,0 +1,383 @@
+import os
+import sys
+import cProfile, pstats
+import pandas
```
So far we have avoided a pandas dependency in the offline code to simplify the number of moving parts, preferring to use vanilla numpy structured arrays or astropy Tables (which aren't necessarily better than Pandas, but just picking that one since we already have an astropy dependency for other reasons).  i.e. can we avoid bringing in pandas as a desimeter dependency?

joesilber replied:
Sure, I'll switch to astropy

sbailey commented on bin/fit_posparams:
```diff
@@ -0,0 +1,383 @@
+# DESI-specific imports
+# path handling is to be improved as I migrate better into github/desimeter
+desimeter_path = os.path.abspath('../py/desimeter/')
+sys.path.append(desimeter_path)
```
The intension for desimeter and other offline code is that users get `desimeter/py` into their `$PYTHONPATH` before running the code, and the code shouldn't manipulate `sys.path` internally.  In particular, this method of finding the library in `../py` relative to `bin/` works when running in a bare git checkout, but won't work in the future when desimeter is installed somewhere (in which case the relationship between the location of the scripts and the python libraries is system-dependent).

joesilber replied:
Ok, will do that

sbailey commented on bin/fit_posparams:
```diff
@@ -0,0 +1,383 @@
+import posparams.fitter as fitter
+import posparams.plotter as plotter
+sys.path.append(fitter.petal_path)
+import petaltransforms
```
Let's discuss this petaltransforms dependency.  So far desimeter has purposefully avoided any dependencies on either online or offline code so that it could be used by either.  At minimum we need to support this without manipulating `sys.path` within the code, but we should discuss this in the broader context of desimeter dependencies.

joesilber replied:
Yeah, also petaltransforms is a bit ugly.
1. I definitely intend to switch over to desimeter.transform.ptl2fp.
2. There is also the more interesting postransforms module, which converts between (t,p) and (x,y). These are not simple transforms. I am debating in my mind forking these conversions into desimeter, in a clean new desimeter-specific module. Would like to discuss pros / cons with you.

sbailey commented on bin/fit_posparams:
```diff
@@ -0,0 +1,383 @@
+# paths
+home_dir = os.getenv('HOME')
+main_dir = os.path.join(home_dir, 'Desktop/posparams')
+data_dir = os.path.join(main_dir, 'fvc_desimeter_v0.2.0_positioners-2') # where to find data files
```
As @julienguy noted, let's specify these locations with input arguments using `argparse` instead of hardcoding paths.

joesilber replied:
yes, will do

sbailey commented on bin/fit_posparams:
```diff
@@ -0,0 +1,383 @@
+main_dir = os.path.join(home_dir, 'Desktop/posparams')
+data_dir = os.path.join(main_dir, 'fvc_desimeter_v0.2.0_positioners-2') # where to find data files
+save_dir = main_dir
+desimeter_data_dir = os.path.join(desimeter_path, 'data')
```
See findfiducials.py for a more robust way of finding the `data` path that will be more platform independent, including when desimeter is installed instead of being used as a bare git checkout.  In short:
```python
from pkg_resources import resource_filename
desimeter_data_dir = resource_filename("desimeter", "data")
```

joesilber replied:
Ok

sbailey commented on bin/fit_posparams:
```diff
@@ -0,0 +1,383 @@
+def last_date_in_sec(frame):
+    return frame['DATE_SEC'].tolist()[-1]
+
+def slice_idxs_for_dates(posid, start=None, final=None):
```
Throughout the code, please add info to the docstrings about the input params, not just the return values.  e.g. here I assume that `posid` is the positioner ID string, but I don't know how to interpret `start` and `final` without reading the code more (restricting the indices of the search? is `final` inclusive or not?).  e.g. like you did in the `process` docstring.

joesilber replied:
Sure

sbailey commented on bin/fit_posparams:
```diff
@@ -0,0 +1,383 @@
+
+if __name__ == '__main__':
+    
+    # define analysis cases spanning the data history
```
Here is where you would add `argparse` option parsing; see other scripts in desimeter/bin for examples.  This has the added benefit of making it safe to run `fit_posparams --help` and get help about input options without having it actually try to do anything.

joesilber replied:
Ok. The purpose of the __main__ check was to ensure multiprocessing safety when running in certain environments (ipython, spyder). I don't see you using this in your scripts like desi_fvc_proc. Do you care about its presence?

sbailey replied:
Including the `if __name__ == '__main__'` wrapper is good practice.  We haven't been strict about this elsewhere because a) we already have a strict separation of what gets imported (stuff in py/desimeter) vs. what gets run from the command line (stuff in bin/) and b) we haven't had people using spyder or other environments where not having it matters.

sbailey commented on py/desimeter/posparams/fitter.py:
```diff
@@ -0,0 +1,182 @@
+# DESI-specific imports
+# path handling is to be improved as I migrate better into github/desimeter
+petal_path = 'C:/Users/joe/Desktop/desi_svn/focalplane/plate_control/trunk/petal/'
+sys.path.append(petal_path)
```
See earlier comments about dependencies and avoiding hardcoded paths.

joesilber replied:
We need to resolve whether I will write a desimeter-specific verison of postransforms. If not, then I need some access to code on the SVN. Is there any standard method in desihub to import SVN code?

joesilber replied:
To close this out --- after discussion we did decide to fork. See PR #56

sbailey commented on py/desimeter/posparams/plotter.py:
```diff
@@ -0,0 +1,191 @@
+# -*- coding: utf-8 -*-
+"""
+Plot time series of positioner parameters.
+"""
```
For the record: it is good that you separated the plotting functions from the algorithmic functions so that the algorithmic functions can be run on a batch node if needed without a X11/GUI connection.

joesilber replied:
Good to know that that is a use case

joesilber commented:
Please see latest posparams. I believe all comments are resolved, and it should work for you at the command line now.

joesilber commented:
Thank you.

>     * test env. variables POSITIONER_LOGS_PATH and FP_SETTINGS_PATH when needed with some help string for how to get the data in DESI SVN if missing.

You will not need this as soon as we merge in the new pos2ptl, and then I remove the old usage of postransforms.py.

> 
>     * `np.diff` does not have argument `prepend` in the numpy versions I use: 1.14.3 and 1.16.4 ( `np.diff` is used in several places in the script `fit_posparams`).
> 
>     * I don't have `numpy.quantile()` function ... hum I see it is in numpy version 1.17

Is it really important that I maintain this backward compatibility? I believe numpy stable version is at 1.18 now...

julienguy commented:
I will let @sbailey  decide on the numpy >=1.7 requirement.

sbailey commented:
Yes, we would like to maintain compatibility with our latest default numpy installation at NERSC, which is 1.16.4 installed last August.  We are open to forcing upgrades if needed for critical bug fixes or super-duper useful new features that would be a pain to otherwise not use, but when workarounds are readily available we prefer to not force people to constantly upgrade their underlying dependencies.  Workarounds here:

* `np.quantile` -> `np.percentile`: equivalent to quantile, but with q in the range [0, 100]. 
* `np.diff(a, prepend=x)`: more irritating to workaround, but something like: `np.concatenate([[x,], np.diff(a)])`

joesilber commented:
- quantile now replaced with percentile
- diff with prepend now replaced with custom function

This should resolve the numpy v1.16 question, hopefully. I did not downgrade my python installation to find out.

joesilber commented:
Per slack conversation, prior to merge, I am going to complete the transition to only desimeter transforms. Thus removing dependency on modules in SVN.

joesilber commented:
postransforms.py is at last expunged, and all tests look like they work. I think it should be ready to merge.

The one place you will still see a "C:/Users/joe..." is in the script `generate_pos2ptl_testdata`. In that case, it is wrapped inside a conditional --- first the script will try to just import postransforms. If that fails, it offers the user the option of trying out the hard-coded path to SVN code (which clearly different users would have to adjust). This script is only for making up values for the unittest of pos2ptl. It's a rare use case.

julienguy commented:
Successfully tested on linux laptop. Good for merging for me.
```
fit_posparams -i M07071.csv -o .
plot_posparams -i 20200423T152820_results_merged.csv -o .
```
and the result, where we see when M07071 started to fail:
![20200423T152820_M07071](https://user-images.githubusercontent.com/5192160/80118287-0ca72000-853d-11ea-9e37-c83e8786f14d.png)

sbailey commented:
Thanks for your patience and all your work wrapping this up.  Merging now.
